pkg/solana/client/cache.go

```diff
@@ -0,0 +1,143 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils"
+
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/config"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/monitor"
+)
+
+type CacheGetter[R any] func(ctx context.Context) (res R, slot uint64, err error)
+
+// Cache is a generic implementation for caching data from the chain
+type Cache[R any] struct {
+	services.StateMachine
+
+	// identifier
+	metricName string
+	Account    solana.PublicKey
+	ChainID    string
+
+	// stored answer
+	resLock sync.RWMutex
+	res     R
+	resTime time.Time
+
+	// dependencies
+	getter CacheGetter[R]
+	cfg    config.Config
+	lggr   logger.Logger
+
+	// polling
+	done   chan struct{}
+	stopCh services.StopChan
+}
+
+func NewCache[R any](metricName string, account solana.PublicKey, chainID string, cfg config.Config, getFunc CacheGetter[R], lggr logger.Logger) *Cache[R] {
+	return &Cache[R]{
+		metricName: metricName,
+		Account:    account,
+		ChainID:    chainID,
+		getter:     getFunc,
+		lggr:       lggr,
+		cfg:        cfg,
+	}
+}
+
+func (c *Cache[R]) Name() string {
+	return c.lggr.Name()
+}
+
+// Start polling
+func (c *Cache[R]) Start(ctx context.Context) error {
+	return c.StartOnce("cache", func() error {
```

Comment:
Qq: Looking at the StateMachine code, particularly https://github.com/smartcontractkit/chainlink-common/blob/4ed45812784df8f781beab8748a5a1b504527990/pkg/services/state.go#L85, I see that the first argument is only used in logs to identify what happened. Would it thus make sense to append the `metricName` to the name here (e.g. `cache_ocr2_median_state` for the `StateCache` and `cache_ocr2_median_transmission` for the `TransmissionsCache`) so that logs better point to what component failed?

Reply:
yes, great suggestion!